### PR TITLE
C++: Mark many non-virtual 'final' methods as 'extern (D)'

### PR DESCRIPTION
**Pull request by ibuclaw**
These are mostly specific to semantic, or changing the AST state. Neither of which is of any relevance to GDC or LDC.

### CODE REVIEW

**Comment by thewilsonator** [comment]
Nice! (cc @kinke fyi)

File: compiler/src/dmd/dclass.d [hunk 1]
```diff
@@ -367,7 +367,7 @@ extern (C++) class ClassDeclaration : AggregateDeclaration
         baseok = Baseok.none;
     }
 
-    final void classError(const(char)* fmt, const(char)* arg)
+    extern (D) final private void classError(const(char)* fmt, const(char)* arg)
```
Comment (thewilsonator) [inline compiler/src/dmd/dclass.d:370]: `src/dmd/dclass.d(370:35)[warn]: Useless final attribute, private functions declared within a class are never virtual`

**Review by thewilsonator — APPROVED** [review]
This otherwise looks good.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
This is great, but if we are on this topic, I was wondering what is the exact interface gdc and ldc are using? Is it documented anywhere? I am asking because we are now trying to use dmd as a library in projects such as d-scanner and dfmt and the fact that the visitor interface is extern(C++) is a royal pain in the pass. Apart from GDC and LDC which are written in C++, other tool creators want to use the full power of D and the extern(C++) is limiting. One such example is if you want to create a templated visitor that inherits from one of the existing visitors you cannot instantiate it with a type that cannot be mapped to C++ (concrete example: you want a format visitor that can print to a generic output stream - stdout, string, LockingTextWritter etc. -). So the question is: are LDC and GDC using the visitor from DMD or are they creating their own visitors? Technically, if AST nodes would remain extern(C++) but the visitors would be extern(D) then the library would improve from a usability standpoint.

**Comment by thewilsonator** [comment]
> what is the exact interface gdc and ldc are using? Is it documented anywhere?

It keep changing because DMD keeps changing. It is not documented, though the `cxxfrontend.cc` is a record of what is used as that gets updated every time there is a regression on it.

> are LDC and GDC using the visitor from DMD 

They are, the codegen and additional semantic analysis subclass the DMD visitors.

File: compiler/src/dmd/expression.d [hunk 2]
```diff
@@ -331,24 +331,24 @@ TemplateDeclaration getFuncTemplateDecl(Dsymbol s) @safe
  * to serve essentially as a Variant that will sit on the stack
  * during CTFE to reduce memory consumption.
  */
-extern (C++) struct UnionExp
+extern (D) struct UnionExp
```
[the remark this answers is not in the document]
Comment (ibuclaw) [reply compiler/src/dmd/expression.d:334]: Hmm, I added `emplace` back in 2019 but it looks like it didn't end up being used in GDC despite implementing our own `Compiler::paintAsType` function (the `UnionExp*` parameter is ignored).

I'll consider just removing `emplace` in another PR.

[the remark this answers is not in the document]
Comment (thewilsonator) [reply compiler/src/dmd/expression.d:334]: FWIW, `git grep UnionExp` return only hits in `dmd/*` for LDC. so it is unused in LDC's C++ codebase.

File: compiler/src/dmd/canthrow.d [hunk 3]
```diff
@@ -55,7 +55,7 @@ enum CT : BE
  *      eSink = if !null, then send error messages to eSink
  * Returns: `CT.exception` or `CT.error` if the expression may throw exceptions.
  */
-extern (C++) /* CT */ BE canThrow(Expression e, FuncDeclaration func, ErrorSink eSink)
+BE /* CT */ canThrow(Expression e, FuncDeclaration func, ErrorSink eSink)
```
[the remark this answers is not in the document]
Comment (thewilsonator) [reply compiler/src/dmd/canthrow.d:58]: This is likewise unused in LDC's C++ codebase.

[the remark this answers is not in the document]
Comment (ibuclaw) [reply compiler/src/dmd/canthrow.d:58]: Added in #13162, based on the final version of the PR, I can't see any good reason for it being added to the C++ frontend tester.

**Comment by ibuclaw** [comment]
> Technically, if AST nodes would remain extern(C++) but the visitors would be extern(D) then the library would improve from a usability standpoint.

Conversely, if the visitors got changed to extern(D) then the library would degrade from a usability standpoint for LDC and GDC.

**Pull request merged**